### PR TITLE
Account for transforms in layer metadata

### LINKED ISSUE #26: Export should take layer transforms into account
I discovered while playing around with a new napari logo that the svg export doesn't include the transforms on the layer. I think the coordinates should be transformed by all layer transforms and line thicknesses should be multiplied by the layer scale, to match what napari is displaying.

See original issue:

https://github.com/jni/new-napari-logo/issues/1

### PR DESCRIPTION
**Pull request by jni**
Closes #26

I haven't accounted for point/line extent in this PR. This was broken before
this and will stay broken after this, but is a somewhat orthogonal problem so
I'd rather leave it to a subsequent PR.

Similarly, we could optimise by pro-actively removing the no-op transforms from
the SVG before writing it out. I can do that here but I would rather focus on
getting it working first and then optimising later.

### CODE REVIEW

**Comment by jni** [comment]
@brisvag @andy-sweet I don't have a test but this is working! As you can see in the commit list it's been a journey. 😅 

Re testing: indeed @andy-sweet I immediately ran into annoying differences in the test svg, like width and height being 128.0 instead of 128 and translate being -0 instead of 0. Maybe as a *test* dependency we could add https://pypi.org/project/svgelements/, which would help us test actual semantic info about the produced svg vs the reference svg, rather than straight-up text?

Maybe the same can be done with the plain xml tree but I have little experience working with it...

File: napari_svg/layer_to_xml.py [hunk 1]
```diff
@@ -44,6 +44,79 @@ def map(self, image):
 }
 
 
+def layer_transforms_to_xml_string(meta):
+    """Get the xml representation[1]_[2]_ of the layer transforms.
+
+    Parameters
+    ----------
+    meta : dict
+        The metadata tuple from the layer.
```
Comment (andy-sweet) [inline napari_svg/layer_to_xml.py:53]: Nit: It's a dictionary, not a tuple, and that's already in the type specifier above, so remove?

```suggestion
        The metadata from the layer.
```

File: napari_svg/_tests/test_write_points_with_attributes-expected.svg [hunk 2]
```diff
@@ -1,4 +1,4 @@
...
\ No newline at end of file
+<svg height="128.0" width="128.0" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><g transform="translate(-0.0 -0.0)"><circle cx="0" cy="0" r="8.0" stroke="rgb(0, 255, 255)" fill="rgb(255, 0, 0)" transform="matrix(1.0 0.0 0.0 1.0 0.0 0.0) translate(0.0 0.0) rotate(0.0) skewY(0.0) scale(1.0 1.0)" stroke-width="0.8" opacity="0.5" /><circle cx="128" cy="0" r="10.0" stroke="rgb(255, 0, 255)" fill="rgb(0, 128, 0)" transform="matrix(1.0 0.0 0.0 1.0 0.0 0.0) translate(0.0 0.0) rotate(0.0) skewY(0.0) scale(1.0 1.0)" stroke-width="5.0" opacity="0.5" /><circle cx="128" cy="128" r="12.0" stroke="rgb(255, 255, 0)" fill="rgb(0, 0, 255)" transform="matrix(1.0 0.0 0.0 1.0 0.0 0.0) translate(0.0 0.0) rotate(0.0) skewY(0.0) scale(1.0 1.0)" stroke-width="12.0" opacity="0.5" /></g></svg>
```
Comment (andy-sweet) [inline napari_svg/_tests/test_write_points_with_attributes-expected.svg:4]: Given there is a top-level group element (in `xml_to_svg`) and the transform for all the circles is the same, I would expect the transform to be defined once there and for each circle to have no transform.

Does that make sense? Or am I misunderstanding something about the SVG spec?

Doesn't necessarily block this PR, but might be worth doing if it's simple and makes the PR smaller.

Comment (andy-sweet) [reply napari_svg/_tests/test_write_points_with_attributes-expected.svg:4]: I just realized we are using that transform to shift everything so the view-box can start at 0, so maybe this isn't as easy as I'd hoped.

**Comment by andy-sweet** [comment]
> e testing: indeed @andy-sweet I immediately ran into annoying differences in the test svg, like width and height being 128.0 instead of 128 and translate being -0 instead of 0. Maybe as a _test_ dependency we could add https://pypi.org/project/svgelements/, which would help us test actual semantic info about the produced svg vs the reference svg, rather than straight-up text?

And all it took was one PR!

The diff itself (as a reviewer of the PR) is very easy to read, especially with the rendered visual diff provided by GitHub.

`svgelements` seems like the kind of thing that I'd want for more extensive testing that also doesn't rely on comparisons with blessed/golden-file checks. Probably best to add an issue where the main work is to define tests that actually check the content of the written SVGs and then mention `svgelements` as a useful option for parsing.

I.e. I don't think we should block this PR.

File: napari_svg/layer_to_xml.py [hunk 3]
```diff
@@ -44,6 +44,79 @@ def map(self, image):
...
+    References
+    ----------
+    .. [1] https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/transform
+    .. [2] https://www.w3.org/TR/css-transforms-1/
+    """
+    scale = meta.get('scale', [1, 1])[::-1]
+    translate = meta.get('translate', [0, 0])[::-1]
+    rotmat = meta.get('rotate', [[1, 0], [0, 1]])
+    rotate = np.degrees(np.arctan2(rotmat[0][1], rotmat[1][1]))
+    # 'shear' in napari specifies the skew along the y-axis in CSS/SVG, but
```
Comment (andy-sweet) [inline napari_svg/layer_to_xml.py:69]: I love the detailed comments here.

**Review by andy-sweet — APPROVED** [review]
I did some very brief manual testing with a points layer, and behavior was as expected after reading this PR. I made some optional suggestions and there are lots of future improvements, but nothing that should block this being merged.

File: napari_svg/_tests/test_write_points_with_attributes-expected.svg (hunk 2 above)
Comment (jni) [reply napari_svg/_tests/test_write_points_with_attributes-expected.svg:4]: yeah so actually you're both right and wrong... There should be one transform per layer. When I looked at the svg changes I thought "huh, I guess he put each point in a different layer. No idea why you would do that, maybe to test multiple layers, but ok." But now I look at the code and I see that there's one transform per shape!

The top level transform is indeed used to make a tight boundary around the data. But looking at the svg spec again it does look like I can nest things with `g` elements: https://developer.mozilla.org/en-US/docs/Web/SVG/Element/g, so a layer should actually go in a g element and the transforms should go in there. I'll have a look shortly.

Comment (jni) [reply napari_svg/_tests/test_write_points_with_attributes-expected.svg:4]: After looking at the code a bit I'm not super confident about the updates, and they will definitely require updating the reference svg, so I'd like to experiment *after* this PR. I'm going to merge and create issues for the remaining work. Thanks for the review @andy-sweet!

**Pull request merged**